docs/integrations/step-patterns.md

```diff
@@ -0,0 +1,326 @@
+guideline is that each step should be hitting _one_ API endpoint at most.
+
+There are a few standard patterns that emerge when designing steps for a
+JupiterOne integration. Here, we assign names to each of these patterns in order
+to build a shared, referenceable grammar to use when developing new JupiterOne
+integrations.
+
+Let's imagine a JupiterOne integration for a device management tool, "Acme".
+This integration ingests the following entities and relationships:
+
+- Entities
+  - `acme_account`
+  - `acme_user`
+  - `acme_group`
+  - `acme_device`
+  - `acme_device_application`
+- Relationships
+  - `acme_account_has_user`
+  - `acme_account_has_group`
+  - `acme_group_has_user`
+  - `acme_user_has_device`
+  - `acme_device_has_application`
+
+In order design the optimal mapping of steps to their entities & relationships,
+you'll need to read the API specification of the target system. For the sake of
+this example, let's assume the API spec is composed of the following endpoints:
+
+- Endpoints
+  - ```
+    GET /account
+    {
+      id: 'account-id',
+      ...
+    }
+    ```
+  - ```
+    GET /users
+    {
+      id: 'user-id',
+      device: { id: 'device-id' },
+      ...
+    }
+    ```
+  - ```
+    GET /groups
+    {
+      id: 'group-id',
+      ...
+    }
+    ```
+  - ```
+    GET /groups/{groupId}/users
+    {
+      id: 'user-id',
+      ...
+    }
+    ```
+  - ```
+    GET /devices
+    {
+      id: 'device-id',
+      ...
+    }
+    ```
+  - ```
+    GET /devices/{deviceId}/applications
+    {
+      id: 'application-id',
+      ...
+    }
+    ```
+
+For this example, based on my API spec, I would ingest these entities and
+relationships in the following steps:
+
+- Steps
+  - **Fetch Account**
+  - **Fetch Users**
+  - **Fetch Groups**
+  - **Fetch Group to User Relationships**
+  - **Fetch Devices**
+  - **Build User to Device Relationships**
+  - **Fetch Applications for Device**
+
+## "Singleton" step pattern
+
+Typically, target systems will have some type of `Account` entity that should be
+pulled in as the integration's root node to the JupiterOne graph. Often, account
+properties are fetched at an endpoint such as `/account` or `/organization`.
+
+The **Singleton** step pattern simply fetches data from that `/account` endpoint
+and creates the root entity. In addition, the `Account` entity is often used to
+build relationships to downstream entities such as `User`s or `Group`s, so we
+typically store this `Account` entity in our job state for easy retrieval later.
+
+The archetype for the **Singleton** pattern:
+
+```typescript
+const fetchAccountStep: Step = {
+  id: 'fetch-account',
+  name: 'Fetch Account',
+  entities: [{ _type: 'acme_account', ... }],
+  relationships: [],
+  dependsOn: [],
+  executionHandler: (context) => {
+    const { jobState, instance } = context;
+    const client = new AcmeClient({ config: instance.config });
```

Comment:
I think we might want to start using a getter pattern for clients instead of creating a new client for each step in order to better control rate and abuse limits.

Reply:
That's a good thought. I'll leave this as is, since the focus is more on everything below this, but I definitely agree with you.